jme3-effects/src/main/resources/Common/MatDefs/Post/ColorContrast15.frag

```diff
@@ -0,0 +1,89 @@
+ *
+ * * Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in the
+ *   documentation and/or other materials provided with the distribution.
+ *
+ * * Neither the name of 'jMonkeyEngine' nor the names of its contributors
+ *   may be used to endorse or promote products derived from this software
+ *   without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+ * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ * LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * Utilized by the ColorContrast.j3md to adjust the color channels contrast by adjusting the input range of the color channels based on
+ * an upper and a lower limit, adjusting the exponent of different color channels and scaling the output values.
+ *
+ * Supports glsl150 and glsl150+ including android GLES glsl300, glsl310 and glsl320.
+ */
+
+#import "Common/ShaderLib/GLSLCompat.glsllib"
+#import "Common/ShaderLib/MultiSample.glsllib"
+
+//constant inputs from java source
+uniform float m_redChannelExponent;
+uniform float m_greenChannelExponent;
+uniform float m_blueChannelExponent;
+
+//final scale values
+uniform float m_redChannelScale;
+uniform float m_greenChannelScale;
+uniform float m_blueChannelScale;
+
+//input range
+uniform float m_lowerLimit;
+uniform float m_upperLimit;
+
+//container for the input from post15.vert
+uniform COLORTEXTURE m_Texture;
+
+//varying input from post15.vert vertex shader
+in vec2 texCoord;
+
+//the output color
+out vec4 fragColor;
+
+void main() {
+    //get the color from a 2d sampler.
+    vec4 color = texture2D(m_Texture, texCoord);
+
+    //apply the color transfer function.
+
+    //1) adjust the channels input range
+    color.rgb = (color.rgb - vec3(m_lowerLimit)) / (vec3(m_upperLimit) - vec3(m_lowerLimit));
+    // limit the chromaticity space into the +ve quadrant
+    color.rgb = abs(color.rgb);
+
+    //2) apply transfer functions on different channels.
+    color.r = pow(color.r, m_redChannelExponent);
+    color.g = pow(color.g, m_greenChannelExponent);
+    color.b = pow(color.b, m_blueChannelExponent);
+
+    //3) apply a final scale factor, between 0.0 and 1.0.
+    color.r = color.r * min(max(m_redChannelScale, 0.0), 1.0);
```

Comment:
Invoking min() and max() in the shader is inefficient. It would be better for Java to ensure the scales passed to the shader are all in the range (0, 1).

Reply:
Although i see no big difference, but I will move the code to the java side for clean purposes.

Reply:
The shader gets executed on every frame; `setScales()` typically gets executed *much* less often. That's the difference.